contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,435 @@
+        bytes32 keyHash,  // Corresponds to a particular offchain job which uses that key for the proofs
+        uint64  subId,
+        uint64  minimumRequestConfirmations,
+        uint64  callbackGasLimit,
+        uint64  numWords  // Desired number of random words
+    )
+    external
+    returns (uint256 requestId)
+    {
+       // One would consider doing a sanity check that the subscription has enough link here,
+       // but:
+       //   1) you'd need to either store a cached price or do a lookup which are both costly
+       //   2) set a fixed min which would be hard to estimate
+       //   3) price could fluctuate anyways between request and response, so it could still fail to fulfill
+       //   4) if it did fail to fulfill, you could simply top up the subscription and re-try and it would succeed
+       require(s_subscriptions[subId].owner != address(0), "invalid subId");
+       require(minimumRequestConfirmations >= s_config.minimumRequestBlockConfirmations, "minconfs too low");
+       bool validConsumer;
+       for (uint16 i = 0; i < s_subscriptions[subId].consumers.length; i++) {
+           if (s_subscriptions[subId].consumers[i] == msg.sender) {
+               validConsumer = true;
+               break;
+           }
+       }
+       require(validConsumer, "invalid consumer");
+       require(s_serviceAgreements[keyHash] != address(0), "must be a registered key");
+       uint256 nonce = s_nonces[keyHash][msg.sender] + 1;
+       uint256 preSeedAndRequestId = uint256(keccak256(abi.encode(keyHash, msg.sender, nonce)));
+       // Min req confirmations not needed as part of fulfillment, leave out of the commitment
+       s_callbacks[preSeedAndRequestId] = keccak256(abi.encodePacked(preSeedAndRequestId, block.number, subId, callbackGasLimit, numWords, msg.sender));
+       emit RandomWordsRequested(keyHash, preSeedAndRequestId, subId, minimumRequestConfirmations, callbackGasLimit, numWords, msg.sender);
+       s_nonces[keyHash][msg.sender] = nonce;
+       return preSeedAndRequestId;
+    }
+
+    function getCallback(uint256 requestId)
+    external
+    view
+    returns (bytes32){
+        return s_callbacks[requestId];
+    }
+
+    // Offsets into fulfillRandomnessRequest's _proof of various values
+    //
+    // Public key. Skips byte array's length prefix.
+    uint256 public constant PUBLIC_KEY_OFFSET = 0x20;
+    // Seed is 7th word in proof, plus word for length, (6+1)*0x20=0xe0
+    uint256 public constant PRESEED_OFFSET = 0xe0;
+
+    function fulfillRandomWords(
+        bytes memory _proof
+    )
+    external
+    {
+        uint256 startGas = gasleft();
+        (bytes32 keyHash, uint256 requestId,
+        uint256 randomness, FulfillmentParams memory fp) = getRandomnessFromProof(_proof);
+
+        uint256[] memory randomWords = new uint256[](fp.numWords);
+        for (uint256 i = 0; i < fp.numWords; i++) {
+            randomWords[i] = uint256(keccak256(abi.encode(randomness, i)));
```

Comment:
```suggestion
            randomWords[i] = uint256(keccak256(randomness + i));
```
Wondering if this would be any cheaper gas wise? Probably not a big difference, but I'de expect it to be a little cheaper.

Reply:
ah we need to encode to get a bytes memory arg for the hash